schema/encoding/jsonschema.go

```diff
@@ -0,0 +1,149 @@
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+
+	"github.com/rs/rest-layer/schema"
+)
+
+type errWriter struct {
+	w   io.Writer
+	err error
+}
+
+func (ew errWriter) writeFormat(format string, a ...interface{}) {
+	if ew.err != nil {
+		return
+	}
+	_, ew.err = fmt.Fprintf(ew.w, format, a...)
+}
+
+func (ew errWriter) writeString(s string) {
+	if ew.err != nil {
+		return
+	}
+	_, ew.err = ew.w.Write([]byte(s))
+}
+
+func (ew errWriter) write(b []byte) {
+	if ew.err != nil {
+		return
+	}
+	_, ew.err = ew.w.Write(b)
+}
+
+// ValidatorToJSONSchema takes a validator and renders to JSON
+func validatorToJSONSchema(w io.Writer, v schema.FieldValidator) (err error) {
+	if v == nil {
+		return nil
+	}
+	ew := errWriter{w: w}
+	switch t := v.(type) {
+	case *schema.String:
+		ew.writeString(`"type": "string"`)
+		if t.Regexp != "" {
+			ew.writeFormat(`, "pattern": %q`, t.Regexp)
+		}
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, fmt.Sprintf("%q", value))
+			}
+			ew.writeFormat(`, "enum": [%s]`, strings.Join(allowed, ", "))
+		}
+		if t.MinLen > 0 {
+			ew.writeFormat(`, "minLength": %s`, strconv.FormatInt(int64(t.MinLen), 10))
+		}
+		if t.MaxLen > 0 {
+			ew.writeFormat(`, "maxLength": %s`, strconv.FormatInt(int64(t.MaxLen), 10))
+		}
+	case *schema.Integer:
+		ew.writeString(`"type": "number"`)
+
+		if len(t.Allowed) > 0 {
+			var allowed []string
+			for _, value := range t.Allowed {
+				allowed = append(allowed, strconv.FormatInt(int64(value), 10))
+			}
+			ew.writeFormat(`, "enum": [%s]`, strings.Join(allowed, ","))
+		}
+		if t.Boundaries != nil {
+			ew.writeFormat(`, "minimum": %s, "maximum": %s`,
+				strconv.FormatFloat(t.Boundaries.Min, 'E', -1, 64),
+				strconv.FormatFloat(t.Boundaries.Max, 'E', -1, 64))
+		}
+	case *schema.Array:
+		ew.writeString(`"type": "array"`)
+		if t.ValuesValidator != nil {
+			ew.writeString(`, "items": `)
+			if ew.err == nil {
+				ew.err = validatorToJSONSchema(w, t.ValuesValidator)
+			}
+		}
+	case *schema.Object:
+		if ew.err == nil {
+			ew.err = schemaToJSONSchema(w, t.Schema)
+		}
+	case *schema.Time:
+		ew.writeString(`"type": "string", "format": "date-time"`)
+	default:
+	}
+	return ew.err
+}
+
+// SchemaToJSONSchema helper
+func schemaToJSONSchema(w io.Writer, s *schema.Schema) (err error) {
+	ew := errWriter{w: w}
+	ew.writeString("{")
+	if s.Description != "" {
+		ew.writeFormat(`"title": %q, `, s.Description)
+	}
+	ew.writeString(`"type": "object", `)
+	ew.writeString(`"properties": {`)
+	var required []string
+	var notFirst bool
+	for key, field := range s.Fields {
+		if notFirst {
+			ew.writeString(", ")
+		}
+		notFirst = true
+		ew.writeFormat("%q: {", key)
+		if field.Description != "" {
+			ew.writeFormat(`"description": %q, `, field.Description)
+		}
+		if field.Required {
+			required = append(required, fmt.Sprintf("%q", key))
+		}
+		if field.ReadOnly {
+			ew.writeString(`"readOnly": {"type": "boolean"}, `)
```

Comment:
I think I understand the confusion. My reasoning for this is each instance
of the output could be read only or not. Your approach says that this field
is always read-only. I think this is where the confusion lies.

If we take the latter interpretation then you are quite correct and I
should change the implementation. Does that make sense? I think the way
schema uses read-only, your interpretation is the correct one. Thanks for
persistently pointing this out.

On Wed, Aug 17, 2016, 07:49 Yan-Fa Li <yanfali@gmail.com> wrote:

> Sorry that doesn't make any sense. It's a specification and not a value.
> Boolean only has two values. What is unclear about that?
>
> It's also in the advanced examples.
>
> ```
> {
>     "id": "http://some.site.somewhere/entry-schema#",
>     "$schema": "http://json-schema.org/draft-04/schema#",
>     "description": "schema for an fstab entry",
>     "type": "object",
>     "required": [ "storage" ],
>     "properties": {
>         "storage": {
>             "type": "object",
>             "oneOf": [
>                 { "$ref": "#/definitions/diskDevice" },
>                 { "$ref": "#/definitions/diskUUID" },
>                 { "$ref": "#/definitions/nfs" },
>                 { "$ref": "#/definitions/tmpfs" }
>             ]
>         },
>         "fstype": {
>             "enum": [ "ext3", "ext4", "btrfs" ]
>         },
>         "options": {
>             "type": "array",
>             "minItems": 1,
>             "items": { "type": "string" },
>             "uniqueItems": true
>         },
>         "readonly": { "type": "boolean" }
>     },
>     "definitions": {
>         "diskDevice": {},
>         "diskUUID": {},
>         "nfs": {},
>         "tmpfs": {}
>     }
> } ```
>
> On Wed, Aug 17, 2016, 02:57 Sindre Røkenes Myren <notifications@github.com>
> wrote:
>
>> In schema/encoding/jsonschema.go
>> <https://github.com/rs/rest-layer/pull/30#discussion_r75094319>:
>>
>> > +	var required []string
>> > +	var notFirst bool
>> > +	for key, field := range s.Fields {
>> > +		if notFirst {
>> > +			ew.writeString(", ")
>> > +		}
>> > +		notFirst = true
>> > +		ew.writeFormat("%q: {", key)
>> > +		if field.Description != "" {
>> > +			ew.writeFormat(`"description": %q, `, field.Description)
>> > +		}
>> > +		if field.Required {
>> > +			required = append(required, fmt.Sprintf("%q", key))
>> > +		}
>> > +		if field.ReadOnly {
>> > +			ew.writeString(`"readOnly": {"type": "boolean"}, `)
>>
>> Change to:
>> ew.writeFormat("readOnly": %q", field.ReadOnly)
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rs/rest-layer/pull/30/files/4ea268f771a0e109990d347deae8aa38ea09b209#r75094319>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ACK5Uw_Emyvbpw4NRjiOIFQT3jiWpCW6ks5qgtr8gaJpZM4JkoxF>
>> .
>>
>